### PR TITLE
Online word2vec test

### PR DESCRIPTION
**Pull request by isohyt**
I was doing the small fix	zachmayer's and rutum's online word2vec code referring to #700.
Moreover, I wrote a test code to check the online word2vec working correctly.
My test strategies are as follows:
1. Check the size of vocabulary in initial training model
2. After online training, checking the size of vocabulary increases
3. each sampling method (hs, neg) change the embedded vector through online training

In Travis-ci, my test is passed, except python 2.6. Do you have any idea on how to solve this?

### CODE REVIEW

**Comment by zachmayer** [comment]
You're getting one test failure on python 2.6: https://travis-ci.org/RaRe-Technologies/gensim/jobs/143268922
```{python}
======================================================================
FAIL: Test that the algorithm is able to add new words to the
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/RaRe-Technologies/gensim/gensim/test/test_word2vec.py", line 121, in testOnlineLearning
    self.assertTrue(numpy.allclose(model_hs.syn0, orig0hs))
AssertionError
```

**Comment by isohyt** [comment]
I think it's just a probabilistic problem so that increase iteration to pass the test.
In my local environment and Travis-ci have passed the test but in this pull request is not :(
(https://travis-ci.org/isohyt/gensim/builds/144057596)

**Comment by isohyt** [comment]
wow! I found much of mistakes in online word2vec code.
I will fix that and write additional test code.

**Comment by isohyt** [comment]
My fixing points are summarized as follows:
1. increase the count of already added vocabulary. In the previous code, only added new vocabulary through online training and not continuing the count previously seen words.
2. check loop. line 1046 does not execute the for loop because len(self.vocab) and len(newsyn0) are same. Thus, I change these architectures.
3. correctly initialize syn1, syn1neg. In online phase, the previous code just changes the last element of syn1, syn1neg so that I add the code to initialize the newly added vector's syn1, syn1neg.

In other words, I made a decision to decrease the checking the similarity between word, 'war' and 'terrorism'. In many cases, these similarities are more than 0.8 ~ 0.9. However, sometimes dramatically fall these similarities' values. I think it occurs because of the small sample size.
To practical evaluation, it's a time to train online w2v on big data.

**Comment by rutum** [comment]
I am going to leave this link here: http://rutumulkar.com/blog/2015/word2vec

This is the original code for Online Word2Vec, which does not face any of the vocab issues that you have discussed.

**Comment by isohyt** [comment]
I mean original online w2v code ignores incrementing the count of registered vocabularies.
In this case, following test cannot be passed.

```
sentences = [
     ['human', 'interface', 'computer'],
     ['survey', 'user', 'computer', 'system', 'response', 'time'],
     ['eps', 'user', 'interface', 'system'],
     ['system', 'human', 'system', 'eps'],
     ['user', 'response', 'time'],
     ['trees'],
     ['graph', 'trees'],
     ['graph', 'minors', 'trees'],
     ['graph', 'minors', 'survey']
 ]

 new_sentences = [
     ['computer', 'artificial', 'intelligence'],
     ['artificial', 'trees'],
      ['human', 'intelligence'],
      ['artificial', 'graph'],
      ['intelligence'],
     ['artificial', 'intelligence', 'system']
  ]

len(list(filter(lambda x: x == 'graph',sum(sentences, [])))) ## == 3
len(list(filter(lambda x: x == 'graph',sum(new_sentences, [])))) ## == 1
...

model = word2vec.Word2Vec(sentences, min_count=0)
self.assertTrue(model.vocab['graph'].count, 3)
model.build_vocab(new_sentences, update=True)
self.assertTrue(model_hs.vocab['graph'].count, 4)
```

We assume the count of word 'graph' will be 4 after online build_vocab, but original code doesn't consider that.
My implementation is from line 607 to line 611 in word2vec.py. (https://github.com/RaRe-Technologies/gensim/pull/778/commits/22dab54c4713a94f22b5f3151908da2becfaf07b)

**Comment by isohyt** [comment]
I've written online word2vec tutorial using 2 different wikipedia dumps.
please, check it out :)

File: gensim/corpora/wikicorpus.py [hunk 1]
```diff
@@ -214,9 +214,9 @@ def extract_pages(f, filter_namespaces=False):
             title = elem.find(title_path).text
             text = elem.find(text_path).text
 
-            ns = elem.find(ns_path).text
-            if filter_namespaces and ns not in filter_namespaces:
-                text = None
+            # ns = elem.find(ns_path).text
```
Comment (tmylk) [inline gensim/corpora/wikicorpus.py]: why is this commented out?

File: gensim/models/doc2vec.py [hunk 2]
```diff
@@ -640,7 +640,7 @@ def reset_from(self, other_model):
         self.docvecs.borrow_from(other_model.docvecs)
         super(Doc2Vec, self).reset_from(other_model)
 
-    def scan_vocab(self, documents, progress_per=10000, trim_rule=None):
+    def scan_vocab(self, documents, update=False, progress_per=10000, trim_rule=None):
```
Comment (tmylk) [inline gensim/models/doc2vec.py]: where is the update parameter used?

Comment (tmylk) [inline gensim/models/doc2vec.py]: where is this update param used?

File: gensim/corpora/wikicorpus.py [hunk 3]
```diff
@@ -214,9 +214,9 @@ def extract_pages(f, filter_namespaces=False):
...
+            # if filter_namespaces and ns not in filter_namespaces:
+            #     text = None
```
Comment (tmylk) [inline gensim/corpora/wikicorpus.py]: why is this commented out?

File: gensim/test/test_word2vec.py [hunk 4]
```diff
@@ -98,6 +107,55 @@ def testLambdaRule(self):
...
+        model_hs.build_vocab(new_sentences, update=True)
+        model_neg.build_vocab(new_sentences, update=True)
+        self.assertTrue(model_hs.vocab['graph'].count, 4)
+        self.assertTrue(model_hs.vocab['artificial'].count, 4)
+        self.assertEqual(len(model_hs.vocab), 14)
+        self.assertEqual(len(model_neg.vocab), 14)
+
+    def onlineSanity(self, model):
+        terro = [l for l in list_corpus if 'terrorism' in l]
+        others = [l for l in list_corpus if 'terrorism' in l]
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: terro looks same as others

File: gensim/test/test_word2vec.py [hunk 5]
```diff
@@ -98,6 +107,55 @@ def testLambdaRule(self):
...
+        model.build_vocab(others)
+        model.train(others)
+        model.build_vocab(terro, update=True)
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: add asserts that no terrorism in old model

File: gensim/corpora/wikicorpus.py (hunk 1 above)
Comment (boqrat) [inline gensim/corpora/wikicorpus.py]: @isohyt  "I've written online word2vec tutorial using 2 different wikipedia dumps"
Can you please provide the link to the tutorial? Thanks!

**Comment by isohyt** [comment]
@boqrat link is [here](https://github.com/isohyt/gensim/blob/8f70e15fbce2ae5446041f91d4d01dc1577ca722/docs/notebooks/online_w2v_tutorial.ipynb)
however, this is beta version doc, so please follow updating

**Comment by boqrat** [comment]
@isohyt Thank you for the link. Have you tried saving the oldmodel, loading it and then continue training? I'm getting a vector of all zeros if I load a previous model, however if I follow your instructions it gives me meaningful results.

**Comment by isohyt** [comment]
@boqrat in my environment, i loaded oldmodel successfully
online_w2v_oldmodel

**Comment by boqrat** [comment]
@isohyt I don't have a problem with loading the model. But when I do load a model, I have issues with retraining the model. Here is a sample of my code. So if I run the following code, vector representation of the new words in inputtext2 will NOT be zero.
`model=gensim.models.Word2Vec()`
`model.build_vocab(inputtext1)`
`model.train(inputtext1)`
`model.build_vocab(inputtext2, update=True)`
`model.train(inputtext2)`

But if I save and load the model, as the following code, vector representation of the new words will be zero.
`model=gensim.models.Word2Vec()`
`model.build_vocab(inputtext1)`
`model.train(inputtext1)`
`model.save("model1")`
`model = gensim.models.Word2Vec.load("model1")`
`model.build_vocab(inputtext2, update=True)`
`model.train(inputtext2)`

Let me know if  you encounter the same problem or if I'm missing anything in my code. Thanks!

**Comment by boqrat** [comment]
I downloaded the latest version and it's working fine now! Thank you for your help:) This is a really useful feature!

File: gensim/corpora/wikicorpus.py (hunk 3 above)
Comment (isohyt) [inline gensim/corpora/wikicorpus.py]: When I try to deal with old wiki dump using WikiCorpus, following error was returned.
comment_out

To merge this feature, other fixing is as follows.
```
try:
    ns = elem.find(ns_path).text
    if filter_namespaces and ns not in filter_namespaces:
            text = None
except AttributeError:
    pass
```

**Comment by isohyt** [comment]
@boqrat Thanks for the announcement. By the way, I will do a small fix later, so that please check my future update :)

**Comment by isohyt** [comment]
@tmylk I've completed fixing the pointed point of code.
After passing your review, I will close this PR and open new PR to gather commit.

**Comment by isohyt** [comment]
In addition, I'll write new tutorial ASAP

**Comment by tmylk** [comment]
@isohyt Please fix merge conflicts and then will be happy to merge this useful feature.

**Comment by isohyt** [comment]
I've updated tutorial to use delta wikipedia and to fix merge confict, this PR is rebased

**Comment by aizsolt** [comment]
Just two quick questions: 1. Are you still using the weight freezing technique that Rutum applied? 2. Is the accuracy still dropping?

File: gensim/test/test_word2vec.py [hunk 6]
```diff
@@ -38,7 +38,7 @@ class LeeCorpus(object):
     def __iter__(self):
         with open(datapath('lee_background.cor')) as f:
```
Comment (piskvorky) [inline gensim/test/test_word2vec.py]: @tmylk Should be `smart_open` (we want to use `smart_open` consistently across gensim).
